examples/pmatch.kl

```diff
@@ -54,47 +38,43 @@
-         (syntax-case more
-           [(cons scrut pats)
-            (pure `(let (x ,scrut) (pmatch-aux ,scrut x ,pats)))])]))]
+      (syntax-case stx ()
+        [(_ ,scrut ,pats ...)
+         (pure `(let [x ,scrut] (pmatch-aux ,scrut x ,pats ...)))]))]
    [pmatch-aux
     (lambda (stx)
-      (syntax-case stx
-        [(list (_ scrut tgt pats))
-         (syntax-case pats
-           [() (pure `(error ',scrut))]
-           [(cons c cs)
-            (syntax-case c
-              [(list (pat rhs))
-               (>>= (else-case pat)
-                    (lambda (e)
-                      (case e
-                        [(just x) (pure `(let (,x ,tgt) ,rhs))]
-                        [(nothing)
-                         (pure `(let (kf (lambda (_) (pmatch-aux ,scrut ,tgt ,cs)))
-                                  (ppat ,tgt ,pat ,rhs (kf 'hi))))])))])])]))]
+      (syntax-case stx (else)
+        [(_ ,scrut ,tgt)
+         (pure `(error ',scrut))]
+        [(_ ,scrut ,tgt (else ,rhs) ,cs ...)
+         (pure rhs)]
+        [(_ ,scrut ,tgt ((else ,x) ,rhs) ,cs ...)
+         (pure `(let [,x ,tgt]
+                  ,rhs))]
+        [(_ ,scrut ,tgt (,pat ,rhs) ,cs ...)
+         (pure `(let (kf (lambda (_) (pmatch-aux ,scrut ,tgt ,cs ...)))
+                  (ppat ,tgt ,pat ,rhs (kf 'hi))))]))]
    [ppat
     (lambda (stx)
-      (syntax-case stx
-        [(list (_ tgt pat ks kf))
-         (syntax-case pat
-           [(ident x)
-            (pure `(let (,x ,tgt) ,ks))]
-           [(cons what args)
-            (>>= (make-temporaries (syntax->list args))
-                 (lambda (temps)
-                   (let (temp-names (map fst temps))
-                     (flet (combine (stxs)
-                             (case stxs
-                               [(nil) ks]
-                               [(:: id-and-stx rest)
-                                `(ppat ,(fst id-and-stx) ,(snd id-and-stx) ,(combine rest) ,kf)]))
-                        (pure `(case ,tgt
-                                 [,(cons-list-syntax what (list->syntax temp-names pat) pat)
-                                  ,(combine temps)]
-                                 [(else other) ,kf]))))))])]))]))
+      (syntax-case stx ()
+        [(_ ,tgt ,x ,ks ,kf)
+         (pure (identifier? x))
+         (pure `(let [,x ,tgt] ,ks))]
+        [(_ ,tgt (,what ,args ...) ,ks ,kf)
+         (>>= (syntax->list args)
+           (lambda (arg-list)
+             (>>= (make-temporaries arg-list)
+               (lambda (temps)
+                 (let (temp-names (map fst temps))
+                   (flet (combine (stxs)
+                           (case stxs
+                             [(nil) ks]
+                             [(:: id-and-stx rest)
+                              `(ppat ,(fst id-and-stx) ,(snd id-and-stx) ,(combine rest) ,kf)]))
+                      (pure `(case ,tgt
+                               [,(cons-list-syntax what (list->syntax temp-names args) args)
```

Comment:
Without the nested calls to `raw-syntax-case`, we no longer have the name `pat` to refer to the `(,what ,args ...)` part of the input. So I used `args` instead. I don't think this matters much?